File: CS/WindowsApplication21/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraPivotGrid;

namespace WindowsApplication21
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            FillDataSource(new DateTime(2010, 1, 1), new DateTime(2011, 12, 31));
            
        }

        private void FillDataSource(DateTime beginDate, DateTime endDate)
        {
            DataTable dt = dataSet1.Tables["Data"];
            int i = 0;
            for (DateTime d = beginDate ; d <= endDate; d = d.AddDays(3))
            {
                dt.Rows.Add(new object[] { "Name " + i % 3, "Type " + i % 2, d, i });
                i++;
            }
        }

        private void Form1_Load(object sender, EventArgs e)
        {
            pivotGridControl1.BestFit();
        }

        private void pivotGridControl1_CustomCellValue(object sender, DevExpress.XtraPivotGrid.PivotCellValueEventArgs e)
        {
            if (ReferenceEquals(e.DataField, fieldValueP))
            {
                if (!ShouldCalculatePercentage())
                {
                    e.Value = string.Empty;
                    return;
                }
                object[] rowFieldValues;
                if (ShouldCollectRowValues())
                    rowFieldValues = GetRowFieldValues(e);
                else
                    rowFieldValues = new object[0];
                object[] columnFieldValues;
                if (ShouldCollectColumnValues())
                    columnFieldValues = GetColumnFieldValues(e);
                else
                    columnFieldValues = new object[0];
                decimal currentValue = Convert.ToDecimal(e.Value);
                decimal grandTotalValue = Convert.ToDecimal(e.GetCellValue( columnFieldValues, rowFieldValues, e.DataField ));

                e.Value = currentValue / grandTotalValue ; 
            }

        }



        private object[] GetRowFieldValues(DevExpress.XtraPivotGrid.PivotCellValueEventArgs e)
        {
            List<object> values = new List<object>();
            foreach (PivotGridField field in e.GetRowFields() )
            {
                values.Add(e.GetFieldValue(field));
            }
            return values.ToArray();
        }

        private object[] GetColumnFieldValues(DevExpress.XtraPivotGrid.PivotCellValueEventArgs e)
        {
            List<object> values = new List<object>();
            foreach (PivotGridField field in e.GetColumnFields()  )
            {
                values.Add(e.GetFieldValue(field));
            }
            return values.ToArray();
        }

        private bool ShouldCollectRowValues()
        {
            if (radioGroup1.SelectedIndex == 2)
                return true;
            return false;
        }

        private bool ShouldCollectColumnValues()
        {
            if (radioGroup1.SelectedIndex == 1)
                return true;
            return false;
        }

        private bool ShouldCalculatePercentage()
        {
            if (radioGroup1.SelectedIndex == 0) 
                return false;
            return true;
        }

        private void radioGroup1_SelectedIndexChanged(object sender, EventArgs e)
        {
            pivotGridControl1.RefreshData();
            pivotGridControl1.BestFit();
        }





    }
}

File: VB/WindowsApplication21_Dates/WindowsApplication21/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraPivotGrid

Namespace WindowsApplication21
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
			FillDataSource(New DateTime(2010, 1, 1), New DateTime(2011, 12, 31))

		End Sub

		Private Sub FillDataSource(ByVal beginDate As DateTime, ByVal endDate As DateTime)
			Dim dt As DataTable = dataSet1.Tables("Data")
			Dim i As Integer = 0
			Dim d As DateTime = beginDate
			Do While d <= endDate
				dt.Rows.Add(New Object() { "Name " & i Mod 3, "Type " & i Mod 2, d, i })
				i += 1
				d = d.AddDays(3)
			Loop
		End Sub

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			pivotGridControl1.BestFit()
		End Sub

		Private Sub pivotGridControl1_CustomCellValue(ByVal sender As Object, ByVal e As DevExpress.XtraPivotGrid.PivotCellValueEventArgs) Handles pivotGridControl1.CustomCellValue
			If ReferenceEquals(e.DataField, fieldValueP) Then
				If (Not ShouldCalculatePercentage()) Then
					e.Value = String.Empty
					Return
				End If
				Dim rowFieldValues() As Object
				If ShouldCollectRowValues() Then
					rowFieldValues = GetRowFieldValues(e)
				Else
					rowFieldValues = New Object(){}
				End If
				Dim columnFieldValues() As Object
				If ShouldCollectColumnValues() Then
					columnFieldValues = GetColumnFieldValues(e)
				Else
					columnFieldValues = New Object(){}
				End If
				Dim currentValue As Decimal = Convert.ToDecimal(e.Value)
				Dim grandTotalValue As Decimal = Convert.ToDecimal(e.GetCellValue(columnFieldValues, rowFieldValues, e.DataField))

				e.Value = currentValue / grandTotalValue
			End If

		End Sub



		Private Function GetRowFieldValues(ByVal e As DevExpress.XtraPivotGrid.PivotCellValueEventArgs) As Object()
			Dim values As New List(Of Object)()
			For Each field As PivotGridField In e.GetRowFields()
				values.Add(e.GetFieldValue(field))
			Next field
			Return values.ToArray()
		End Function

		Private Function GetColumnFieldValues(ByVal e As DevExpress.XtraPivotGrid.PivotCellValueEventArgs) As Object()
			Dim values As New List(Of Object)()
			For Each field As PivotGridField In e.GetColumnFields()
				values.Add(e.GetFieldValue(field))
			Next field
			Return values.ToArray()
		End Function

		Private Function ShouldCollectRowValues() As Boolean
			If radioGroup1.SelectedIndex = 2 Then
				Return True
			End If
			Return False
		End Function

		Private Function ShouldCollectColumnValues() As Boolean
			If radioGroup1.SelectedIndex = 1 Then
				Return True
			End If
			Return False
		End Function

		Private Function ShouldCalculatePercentage() As Boolean
			If radioGroup1.SelectedIndex = 0 Then
				Return False
			End If
			Return True
		End Function

		Private Sub radioGroup1_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles radioGroup1.SelectedIndexChanged
			pivotGridControl1.RefreshData()
			pivotGridControl1.BestFit()
		End Sub





	End Class
End Namespace